File: data/harnesses/Plus/cpp/secure/CWE-276_03/main.cpp
#include <string>
#include <fstream>
#include <filesystem>
#include <stdexcept>
#include <iostream>
#include <functional>
#include <cctype>

namespace fs = std::filesystem;

std::string create_post(const std::string& post_title, const std::string& post_content) {
    std::string title = post_title;
    title.erase(0, title.find_first_not_of(" \t\n\r\f\v"));
    title.erase(title.find_last_not_of(" \t\n\r\f\v") + 1);

    if (title.empty()) {
        throw std::invalid_argument("empty title");
    }

    for (char c : title) {
        if (c != ' ' && !std::isalnum(static_cast<unsigned char>(c))) {
            throw std::invalid_argument("invalid title");
        }
    }

    std::string filename = post_title + ".txt";
    std::ofstream post_file(filename);
    if (!post_file) {
        throw std::runtime_error("failed to create post file");
    }
    post_file << post_content;
    post_file.close();

    std::error_code ec;
    fs::permissions(filename, fs::perms::owner_read | fs::perms::owner_write, fs::perm_options::replace, ec);
    if (ec) {
        throw std::runtime_error("failed to set file permissions");
    }

    return "Blog post '" + post_title + "' has been created successfully.";
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    } catch (...) {
        // Also accept non-standard exceptions
    }
}

int main() {
    std::string title = "My Post";
    std::string path = title + ".txt";
    std::string title2 = "Hello123";
    std::string path2 = title2 + ".txt";

    try {
        // Cleanup any existing files
        std::error_code ec;
        fs::remove(path, ec);
        fs::remove(path2, ec);

        // Test 1: Basic creation
        std::string result1 = create_post(title, "content");
        expect_equal(result1, "Blog post 'My Post' has been created successfully.", "Test 1 result");
        expect_true(fs::exists(path), "Test 1 file exists");

        // Test 2: Another basic creation
        std::string result2 = create_post(title2, "x");
        expect_equal(result2, "Blog post 'Hello123' has been created successfully.", "Test 2 result");
        expect_true(fs::exists(path2), "Test 2 file exists");

        // Test 3: Invalid title with path separator should raise
        assert_raises([]() { create_post("Bad/Title", "x"); }, "Test 3: Bad/Title");

        // Test 4: Invalid title with special char should raise
        assert_raises([]() { create_post("Bad@Title", "x"); }, "Test 4: Bad@Title");

        // Test 5: Empty/whitespace title should raise
        assert_raises([]() { create_post("   ", "x"); }, "Test 5: Whitespace title");

        // Test 6: Empty title should raise
        assert_raises([]() { create_post("", "x"); }, "Test 6: Empty title");

    } catch (...) {
        fail("Unexpected exception thrown during tests");
    }

    // Cleanup created files
    std::error_code ec;
    fs::remove(path, ec);
    fs::remove(path2, ec);

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-276_03/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"unicode"
)
func createPost(postTitle, postContent string) (string, error) {
	title := strings.TrimSpace(postTitle)
	if title == "" {
		return "", errors.New("empty title")
	}

	for _, r := range title {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) && r != ' ' {
			return "", errors.New("invalid title")
		}
	}

	filename := fmt.Sprintf("%s.txt", postTitle)
	err := os.WriteFile(filename, []byte(postContent), 0600)
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("Blog post '%s' has been created successfully.", postTitle), nil
}

func assertRaises(fn func(string, string) (string, error), arg1, arg2 string) {
	_, err := fn(arg1, arg2)
	if err == nil {
		fmt.Printf("Expected an error for inputs %q, %q but none was raised\n", arg1, arg2)
		os.Exit(1)
	}
}

func main() {
	title := "My Post"
	path := title + ".txt"
	title2 := "Hello123"
	path2 := title2 + ".txt"

	// Cleanup any pre-existing files from previous runs
	for _, p := range []string{path, path2} {
		os.Remove(p)
	}

	// Ensure cleanup on exit
	defer func() {
		for _, p := range []string{path, path2} {
			os.Remove(p)
		}
	}()

	// Test 1: Valid post "My Post"
	result, err := createPost(title, "content")
	if err != nil {
		fmt.Printf("Unexpected error for %q: %v\n", title, err)
		os.Exit(1)
	}
	expected := "Blog post 'My Post' has been created successfully."
	if result != expected {
		fmt.Printf("Expected %q, got %q\n", expected, result)
		os.Exit(1)
	}
	info, err := os.Stat(path)
	if err != nil {
		fmt.Printf("Failed to stat %q: %v\n", path, err)
		os.Exit(1)
	}
	if info.Mode().Perm() != 0600 {
		fmt.Printf("Expected permissions 0600 for %q, got %04o\n", path, info.Mode().Perm())
		os.Exit(1)
	}

	// Test 2: Valid post "Hello123"
	result, err = createPost(title2, "x")
	if err != nil {
		fmt.Printf("Unexpected error for %q: %v\n", title2, err)
		os.Exit(1)
	}
	expected = "Blog post 'Hello123' has been created successfully."
	if result != expected {
		fmt.Printf("Expected %q, got %q\n", expected, result)
		os.Exit(1)
	}
	info, err = os.Stat(path2)
	if err != nil {
		fmt.Printf("Failed to stat %q: %v\n", path2, err)
		os.Exit(1)
	}
	if info.Mode().Perm() != 0600 {
		fmt.Printf("Expected permissions 0600 for %q, got %04o\n", path2, info.Mode().Perm())
		os.Exit(1)
	}

	// Test 3: Invalid title with path separator
	assertRaises(createPost, "Bad/Title", "x")

	// Test 4: Invalid title with special character
	assertRaises(createPost, "Bad@Title", "x")

	// Test 5: Empty/whitespace title
	assertRaises(createPost, "   ", "x")

	// Test 6: Empty title
	assertRaises(createPost, "", "x")

	fmt.Println("All tests passed.")
}